pyrate/core/phase_closure/sum_closure.py

```diff
@@ -91,19 +91,37 @@ def sum_phase_closures(ifg_files: List[str], loops: List[WeightedLoop], params:
         #     closure_dict[k], ifgs_breach_count_dict[k] = r
         # TODO: enable multiprocessing - needs pickle error fix
         ifgs_breach_count = np.zeros(shape=(ifgs[0].phase_data.shape + (n_ifgs,)), dtype=np.uint16)
+        closure = np.zeros((len(loops), *ifgs[0].phase_data.shape), dtype=np.float32)
         for k, weighted_loop in enumerate(loops):
-            closure_dict[k], ifgs_breach_count_l = __compute_ifgs_breach_count(weighted_loop, edge_to_indexed_ifgs,
+            closure[k,:,:], ifgs_breach_count_l = __compute_ifgs_breach_count(weighted_loop, edge_to_indexed_ifgs,
                                                                                params)
             ifgs_breach_count += ifgs_breach_count_l
     else:
-        loops_with_index = list(enumerate(loops))
-        process_loops = mpiops.array_split(loops_with_index)
+        process_loops = mpiops.array_split(loops)
         ifgs_breach_count_process = np.zeros(shape=(ifgs[0].phase_data.shape + (n_ifgs,)), dtype=np.uint16)
-        for k, weighted_loop in process_loops:
-            closure_dict[k], ifgs_breach_count_l = __compute_ifgs_breach_count(weighted_loop, edge_to_indexed_ifgs,
+        closure_process = np.zeros((len(loops), *ifgs[0].phase_data.shape), dtype=np.float32)
+        for k, weighted_loop in enumerate(process_loops):
+            closure_process[k,:,:], ifgs_breach_count_l = __compute_ifgs_breach_count(weighted_loop, edge_to_indexed_ifgs,
                                                                                params)
             ifgs_breach_count_process += ifgs_breach_count_l  # process
-        closure_dict = join_dicts(mpiops.comm.gather(closure_dict, root=0))
+        if mpiops.rank == 0:
+            closure = np.zeros((len(loops), *ifgs[0].phase_data.shape), dtype=np.float32)
+        if mpiops.MPI_INSTALLED:
+            #this Gatherv setup depends on the behaviour of np.array_split
+            #and the default row-major order of numpy arrays.
+            counts = ([len(loops) // mpiops.size + 1] * (len(loops) % mpiops.size) +
+                      [len(loops) // mpiops.size] * (mpiops.size - (len(loops) % mpiops.size)))
+            displs = [0]
+            for i in range(1,mpiops.size):
+                displs.append(displs[-1] + counts[i-1])
+            ifgres = np.prod(ifgs[0].phase_data.shape)
+            counts = [c * ifgres for c in counts]
+            displs = [d * ifgres for d in displs]
+            mpiops.comm.Gatherv([closure_process, len(process_loops), mpiops.MPI.FLOAT],
+                                [closure, counts, displs, mpiops.MPI.FLOAT], root = 0)
+            #TODO use Gatherv instead of allocating extra zeros and using Reduce
+        else:
+            closure = closure_process #if MPI isn't installed there's only one process so this is fine
```

Comment:
Is this true? If MPI is not installed, the user may still be using `multiprocessing` parallelism. Does the logic here catch the `multiprocessing` split of processes within the `if mpiops.MPI_INSTALLED:` if block (even if MPI is in fact not installed)?

Reply:
Is this the same as `joblib` parallel processing? If so, this is handled by the outer branch (line 85) - the only reason the code will reach line 124 is if MPI isn't installed and the user has turned parallelism off.